File: android_mobile/src/test/java/com/alexstyl/specialdates/events/DateTest.java
package com.alexstyl.specialdates.events;

import com.alexstyl.specialdates.date.Date;
import com.alexstyl.specialdates.date.DateComparator;
import com.alexstyl.specialdates.date.MonthInt;

import org.junit.Test;

import static com.alexstyl.specialdates.date.Months.*;
import static org.fest.assertions.api.Assertions.assertThat;

public class DateTest {

    private static final DateComparator comparator = DateComparator.INSTANCE;

    private static final int DAY = 5;
    private static final
    @MonthInt
    int MONTH = 10;

    private static final Date ANY_DATE = Date.Companion.on(DAY, MONTH, 1990);

    @Test
    public void whenAddingADayOfMonth_thenIncreaseNormally() {

        Date after = ANY_DATE.addDay(1);
        int expectedDay = DAY + 1;

        assertThat(after.getDayOfMonth()).isEqualTo(expectedDay);
    }

    @Test
    public void givenAEndOfTheYearDate_whenAddingOneDay_thenTheFirstDayOftheNextYearIsReturned() {
        Date lastDayOfYear = Date.Companion.on(31, DECEMBER, 1990);
        Date firstDayOfNextYear = lastDayOfYear.addDay(1);

        assertThat(firstDayOfNextYear.getMonth()).isEqualTo(1);
        assertThat(firstDayOfNextYear.getDayOfMonth()).isEqualTo(1);
        int nextYear = lastDayOfYear.getYear() + 1;
        assertThat(firstDayOfNextYear.getYear()).isEqualTo(nextYear);
    }

    @Test
    public void testOneDayAhead() {
        Date firstDayOfYear = Date.Companion.on(1, JANUARY, 1990);
        Date secondDayOfYear = Date.Companion.on(2, JANUARY, 1990);
        assertThat(firstDayOfYear.daysDifferenceTo(secondDayOfYear)).isEqualTo(1);
    }

    @Test
    public void testOneYearAhead() {
        Date firstDayOfYear = Date.Companion.on(1, JANUARY, 1990);
        Date secondDayOfYear = Date.Companion.on(1, JANUARY, 1991);
        assertThat(firstDayOfYear.daysDifferenceTo(secondDayOfYear)).isEqualTo(365);
    }

    @Test
    public void testTwoYearAhead() {
        Date firstDayOfYear = Date.Companion.on(1, JANUARY, 1990);
        Date secondDayOfYear = Date.Companion.on(1, JANUARY, 1992);
        assertThat(firstDayOfYear.daysDifferenceTo(secondDayOfYear)).isEqualTo(365 * 2);
    }

    @Test
    public void compareFutureDayDate() {
        int result = comparator.compare(Date.Companion.on(1, JANUARY, 1990), Date.Companion.on(2, JANUARY, 1990));
        assertThat(result).isNegative();
    }

    @Test
    public void compareSameDayDate() {
        int result = comparator.compare(Date.Companion.on(1, JANUARY, 1990), Date.Companion.on(1, JANUARY, 1990));
        assertThat(result).isZero();
    }

    @Test
    public void comparePastDayDate() {
        int result = comparator.compare(Date.Companion.on(1, JANUARY, 1990), Date.Companion.on(1, JANUARY, 1980));
        assertThat(result).isPositive();
    }

    @Test
    public void compareFutureDate() {
        int result = comparator.compare(Date.Companion.on(1, JANUARY, 1990), Date.Companion.on(2, JANUARY));
        assertThat(result).isNegative();
    }

    @Test
    public void compareSameDate() {
        int result = comparator.compare(Date.Companion.on(1, JANUARY, 1990), Date.Companion.on(1, JANUARY));
        assertThat(result).isZero();
    }

    @Test
    public void comparePastDate() {
        int result = comparator.compare(Date.Companion.on(2, JANUARY, 1990), Date.Companion.on(1, JANUARY));
        assertThat(result).isPositive();
    }

    @Test
    public void whenComparingToSameDate_thenTheyAreEqual() {
        Date firstDate = Date.Companion.on(16, APRIL, 1991);
        Date secondDate = Date.Companion.on(16, APRIL, 1991);

        assertThat(firstDate.equals(secondDate)).isTrue();
    }

    @Test
    public void whenComparingToDateWithDifferentYear_thenTheyAreNotEqual() {
        Date firstDate = Date.Companion.on(16, APRIL, 1991);
        Date secondDate = Date.Companion.on(16, APRIL, 1987);

        assertThat(firstDate.equals(secondDate)).isFalse();
    }

    @Test(expected = IllegalArgumentException.class)
    public void throwsException_whenInvalidDateIsCreated() {
        Date.Companion.on(31, FEBRUARY, 1991);
    }
}

File: memento/src/test/java/com/alexstyl/specialdates/events/DateTest.kt
package com.alexstyl.specialdates.events

import com.alexstyl.specialdates.date.DateComparator
import com.alexstyl.specialdates.date.MonthInt
import com.alexstyl.specialdates.date.Months.APRIL
import com.alexstyl.specialdates.date.Months.DECEMBER
import com.alexstyl.specialdates.date.Months.FEBRUARY
import com.alexstyl.specialdates.date.Months.JANUARY
import com.alexstyl.specialdates.date.dateOn
import org.fest.assertions.api.Assertions.assertThat
import org.junit.Test

class DateTest {

    @Test
    fun whenAddingADayOfMonth_thenIncreaseNormally() {

        val after = ANY_DATE.addDay(1)
        val expectedDay = DAY + 1

        assertThat(after.dayOfMonth).isEqualTo(expectedDay)
    }

    @Test
    fun givenAEndOfTheYearDate_whenAddingOneDay_thenTheFirstDayOfTheNextYearIsReturned() {
        val lastDayOfYear = dateOn(31, DECEMBER, 1990)
        val firstDayOfNextYear = lastDayOfYear.addDay(1)

        assertThat(firstDayOfNextYear.month).isEqualTo(1)
        assertThat(firstDayOfNextYear.dayOfMonth).isEqualTo(1)
        val nextYear = lastDayOfYear.year!! + 1
        assertThat(firstDayOfNextYear.year).isEqualTo(nextYear)
    }

    @Test
    fun givenDateWithShortMonthAndNoYearSpecified_thenReturn29Days() {
        val date = dateOn(1, FEBRUARY)

        assertThat(date.daysInCurrentMonth).isEqualTo(29)
    }

    @Test
    fun givenDateWithShortMonthAndCommonYear_thenReturn28Days() {
        val date = dateOn(1, FEBRUARY, 2018)

        assertThat(date.daysInCurrentMonth).isEqualTo(28)
    }

    @Test
    fun givenDateWithShortMonthAndLeapYear_thenReturn29Days() {
        val date = dateOn(1, FEBRUARY, 2020)

        assertThat(date.daysInCurrentMonth).isEqualTo(29)
    }

    @Test
    fun testOneDayAhead() {
        val firstDayOfYear = dateOn(1, JANUARY, 1990)
        val secondDayOfYear = dateOn(2, JANUARY, 1990)
        assertThat(firstDayOfYear.daysDifferenceTo(secondDayOfYear)).isEqualTo(1)
    }

    @Test
    fun testOneYearAhead() {
        val firstDayOfYear = dateOn(1, JANUARY, 1990)
        val secondDayOfYear = dateOn(1, JANUARY, 1991)
        assertThat(firstDayOfYear.daysDifferenceTo(secondDayOfYear)).isEqualTo(365)
    }

    @Test
    fun testTwoYearAhead() {
        val firstDayOfYear = dateOn(1, JANUARY, 1990)
        val secondDayOfYear = dateOn(1, JANUARY, 1992)
        assertThat(firstDayOfYear.daysDifferenceTo(secondDayOfYear)).isEqualTo(365 * 2)
    }

    @Test
    fun compareFutureDayDate() {
        val result = comparator.compare(dateOn(1, JANUARY, 1990), dateOn(2, JANUARY, 1990))
        assertThat(result).isNegative
    }

    @Test
    fun compareSameDayDate() {
        val result = comparator.compare(dateOn(1, JANUARY, 1990), dateOn(1, JANUARY, 1990))
        assertThat(result).isZero
    }

    @Test
    fun comparePastDayDate() {
        val result = comparator.compare(dateOn(1, JANUARY, 1990), dateOn(1, JANUARY, 1980))
        assertThat(result).isPositive
    }

    @Test
    fun compareFutureDate() {
        val result = comparator.compare(dateOn(1, JANUARY, 1990), dateOn(2, JANUARY))
        assertThat(result).isNegative
    }

    @Test
    fun compareSameDate() {
        val result = comparator.compare(dateOn(1, JANUARY, 1990), dateOn(1, JANUARY))
        assertThat(result).isZero
    }

    @Test
    fun comparePastDate() {
        val result = comparator.compare(dateOn(2, JANUARY, 1990), dateOn(1, JANUARY))
        assertThat(result).isPositive
    }

    @Test
    fun whenComparingToSameDate_thenTheyAreEqual() {
        val firstDate = dateOn(16, APRIL, 1991)
        val secondDate = dateOn(16, APRIL, 1991)

        assertThat(firstDate == secondDate).isTrue
    }

    @Test
    fun whenComparingToDateWithDifferentYear_thenTheyAreNotEqual() {
        val firstDate = dateOn(16, APRIL, 1991)
        val secondDate = dateOn(16, APRIL, 1987)

        assertThat(firstDate == secondDate).isFalse
    }

    @Test(expected = IllegalArgumentException::class)
    fun throwsException_whenInvalidDateIsCreated() {
        dateOn(31, FEBRUARY, 1991)
    }

    companion object {

        private val comparator = DateComparator.INSTANCE

        private val DAY = 5
        @MonthInt
        private val MONTH = 10

        private val ANY_DATE = dateOn(DAY, MONTH, 1990)
    }
}
